### ISSUE TITLE
Winforms process not stopping correctly when using WebView2

### ISSUE THREAD

**Issue by Mr-Technician** [dotnet/maui#7997]
### .NET version

6.0.5

### Did it work in .NET Framework?

Not relevant to WebView2.

### Issue description

I have created a Winforms app using Blazor through WebView2. This works great until I try to close the app.

While running, all of the processes are contained under the parent under the Apps section of task manager:
![image](https://user-images.githubusercontent.com/26885142/172871923-ec78f60d-77dc-41e7-aa14-0970d2517192.png)

After closing the app with the close button, I see this in the task manager under the background process section:
![image](https://user-images.githubusercontent.com/26885142/172872696-5bd80396-069e-4fd4-bfcf-f3c6c220a87c.png)
The webview processes also remain in the background:
![image](https://user-images.githubusercontent.com/26885142/172872866-b62a3afe-553b-4f43-9892-4dc4cd50c3e4.png)

If I manually kill the app process at this stage, the WebView processes also stop.

My current fix is the following:

```csharp
private void Form1_FormClosing(object? sender, FormClosingEventArgs e)
        {
             Environment.Exit(0); //calling Exit kills the background process associated with WebView2
        }
```

This kills the background process, which in turn kills the WebView2 processes.

### Steps to reproduce

Create a new Winforms app and create a blazor webview2 app as per: https://docs.microsoft.com/en-us/aspnet/core/blazor/hybrid/tutorials/windows-forms?view=aspnetcore-6.0

Publish the app and run it, ideally without Visual Studio running to make sure nothing else is running WebView2. Close the app, and check to see if the WebView2 processes are still running.

I have observed this issue on a testing server with the published build, not when running in debug mode from Visual Studio.

**Comment by Eilon** [comment]
Moved to MAUI repo so we can investigate as a potential Blazor Hybrid issue.

**Comment by Eilon** [comment]
Hi @janseris , where we placed BlazorWebView was a trade-off and we felt that keeping it in the MAUI repo was the best overall option. The primary scenario was the .NET MAUI BlazorWebView, and also BlazorWebView MAUI+WPF+WinForms depends on nearly _everything_ in .NET, so the MAUI repo fits those requirements. We _could_ place it in a separate repo, but the infrastructure cost to doing that would be enormous, which means that we would have less engineering time to spend on building the features.

**Comment by mkArtakMSFT** [comment]
@MackinnonBuck can you please look into this one? Thanks!

**Comment by MackinnonBuck** [comment]
Hm, I'm not able to repro this behavior. All WebView2-related child process stop when the parent process stops (no matter how I close the Windows Forms app).

@Mr-Technician, could you try using the latest .NET 6 SDK and `Microsoft.AspNetCore.Components.WebView.WindowsForms` package (I tried with 6.0.540) and see if you're still experiencing the issue?

If the issue persists, would you be able to provide a minimal repro project hosted as a public GitHub repository? There's a chance this issue just doesn't show up when creating a Windows Forms Blazor Hybrid app from the docs. Thanks!

**Comment by Mr-Technician** [comment]
@MackinnonBuck So far the issue remains using the latest .NET 6 SDK. I am preparing a minimal repo now.

**Comment by Eilon** [comment]
I suspect the core problem is whatever is keeping your app's main process alive, and thus the associated WebView2 processes also remain alive.

Can you see which thread(s) are running in your process? Just run it normally, and then when you get into that state, you can attach the VS debugger to that process. That will let you see which threads are doing what, and could help us determine the root cause.

The only similar issue I could find in the WebView2 repo is this one: https://github.com/MicrosoftEdge/WebView2Feedback/issues/1424. But even that doesn't seem to indicate any known bug in WebView2.

**Comment by Mr-Technician** [comment]
Here you go:
![image](https://user-images.githubusercontent.com/26885142/192000881-4a619db7-f2f6-4d56-b328-0346140d569d.png)

For comparison, here is the output from my attempted minimal repo, where i was not able to reproduce the issue:
![image](https://user-images.githubusercontent.com/26885142/192002242-f5bc8d86-702c-41dc-bcdb-2cd8a00faa2d.png)

The app with the issue uses the Graph API to get and modify applications in Azure.

**Comment by Eilon** [comment]
@Mr-Technician I'm hoping it's possible for you to try to narrow things down. If you change your app to not make any API calls, does that change anything? Also, can you confirm exactly how the app is being closed? Is the default "X" button in the window's caption bar being clicked, or via some other technique?

**Comment by Mr-Technician** [comment]
@Eilon No luck so far, I commented out my <App /> component so that nothing loads but the basic structure of the app (zero API calls or anything special) and the process + WebView2 processes linger after closing using the default "X" button.

**Comment by Mr-Technician** [comment]
I take it back, I think I found the issue:

I commented out my entire <App /> component and the issue was resolved. Here is my minimal App.razor:

```csharp
<MudThemeProvider />
<MudDialogProvider />
<MudSnackbarProvider />

<MudContainer MaxWidth="MaxWidth.Large">
    @*<Home />*@
</MudContainer>
```

The Mud components are from MudBlazor: https://github.com/MudBlazor/MudBlazor It would seem that something with one of the Providers listed above causes Blazor to hang.

**Comment by Mr-Technician** [comment]
@Eilon I have prepared a minimal repo here, complete with simple reproduction steps: https://github.com/Mr-Technician/BlazorWebview2Minimal

**Comment by MackinnonBuck** [comment]
I was able to narrow down the cause for this bug to the following: https://github.com/dotnet/maui/blob/6ee55222aada4476ae320dcee49818bc498dd6d7/src/BlazorWebView/src/WindowsForms/BlazorWebView.cs#L247-L261

There seem to be a couple problems here:
1. The comment implies that Blazor component disposal logic will complete before `base.Dispose()` gets executed. This is only true for _synchronous_ component disposal logic, because [`PageContext.DisposeAsync()`](https://github.com/dotnet/aspnetcore/blob/41bed3fc073a11f91db34ca16ea42570ff2c5a78/src/Components/WebView/WebView/src/PageContext.cs#L53) calls the synchronous `Renderer.Dispose()`, which does a fire-and-forget for async disposable components. This means we thought we were waiting for components to dispose asynchronously, but this isn't actually the case (and I've created a repro verifying this, and it can cause nasty exceptions to be thrown). The only thing we're blocking on here is the disposal of the `AsyncServiceScope` for the web view. This brings us to the actual cause of the bug...
2. If a service implements `IAsyncDisposable` and does something actually async (even a `Task.Yield()`), it will result in a deadlock because we're blocking synchronously on line 255 above.

In this specific case, the MudBlazor [`MudPopoverService.DisposeAsync()`](https://github.com/MudBlazor/MudBlazor/blob/master/src/MudBlazor/Components/Popover/MudPopoverService.cs#L196) method attempts to perform JS interop when the service disposes, which causes a deadlock for the reason described above.

So I guess my questions are the following:
1. Why does `PageContext` call `Renderer.Dispose()` instead of `Renderer.DisposeAsync()`? It seems that the issue I described in point 1 above would apply to all Blazor Hybrid platforms. Is this something we should change?
2. Assuming we keep things the way they are, should we fire-and-forget the call to `_webViewManager.DisposeAsync()` rather than synchronously blocking? Given that we're already essentially doing that for Blazor component disposal, I don't see much of an advantage to block on disposal of the `AsyncServiceScope` (especially if it causes a deadlock). We could probably find another way to bubble up any exceptions that get thrown.

cc @Eilon @javiercn
